Automattic/woocommerce-payments
# RPP: Add Psalm exception handling to src

Fixes #6953

#### Changes proposed in this Pull Request

https://github.com/Automattic/woocommerce-payments/pull/6908 introduced exceptions within the new `src` directory, which triggered a discussion around exceptions and catching them.

@marcinbot suggested Psalm's `checkForThrowsDocblock` configuration, which turned out to use full analysis, and report uncaught exceptions from nested methods, not just one thrown in the current method. This PR implements it.

Details:

- Configuration is added to `psalm.xml` to check for exceptions.
- `includes` is excluded from the new check, as there are 287 issues to fix there (I was actually expecting more).
- Even though `includes` is excluded, it was preventing Psalm from working. Turns out that `WCPay\Exceptions\Order_Not_Found_Exception` was used in a few places without `use` statements for it. Those are fixed here.
- The `ExampleService` and `ExampleServiceWithDependencies` classes have exceptions and handlers now, to test the configuration.

#### Testing instructions

1. All checks here should be green.
2. Remove the `@throws` tag and the `try - catch` statement from `src/Internal/Service/ExampleServiceWithDependencies.php`.
3. Execute `npm run psalm` locally, and observe the errors.

-------------------

- [x] Covered with tests (no tests really, this is not PHP)
